src/arithmetic/canonical.cc

```diff
@@ -118,11 +119,12 @@ struct BinaryExpr {
 template<typename T>
 inline Expr Binary_(const T* op,
                     const Expr& e,
-                    Expr a, Expr b) {
+                    Expr a, Expr b,
+                    const Map<Var, Range>& vrange) {
   if (a.same_as(op->a) && b.same_as(op->b)) {
     return e;
   } else {
-    return T::make(a, b);
+    return Simplify(T::make(a, b), vrange);
```

Comment:
Theoretically, all (not only binary) operators can use var range information. But this code only handles binary ops.

Reply:
For now let us not mix Simplify and CanonicalSimplify in recursive calls, as it will cost copy of the vrange info and increase memory complexity

Reply:
To resolve concerns regarding increased memory usage: why don't we pass the range by reference instead of by value? would that resolve your concern?
There is obviously the extra time that we need to pay for calling Simplify inside CanonicalSimplify. But that's exactly how we get the benefit of the rules in Simplify. Is there any other way?

Reply:
We can simply call Simplify after CanonicalSimplify, which should give you similar effect for now.

Reply:
My understanding is that `reduce` is not part of Halide IR (it's a TVM op). And Simplify is supposed to be called only on expressions with only Halide IR ops. Moreover, CanonicalSimplify is supposed to be called on TVM expressions (that may have non-Halide IR ops). 
Specifically, the expression I was interested in simplifying has a `reduce` as top operator and that's (one of the reasons) why I didn't run Simplify on it.

Reply:
I have update CanonicalSimplify in https://github.com/dmlc/tvm/pull/676 so it works for your case.